src/test/java/org/springframework/integration/kafka/listener/KafkaMessageDrivenChannelAdapterTests.java

```diff
@@ -220,4 +222,131 @@ public boolean send(Message<?> message, long timeout) {
 			assertThat(metadataStore.get(offsetManager.generateKey(readPartition)), equalTo(String.valueOf(20)));
 		}
 	}
+
+	@Test
+	@SuppressWarnings("serial")
+	public void testShutdownWaitsForContainerToStop() throws Exception {
+		createTopic(TEST_TOPIC, 5, 1, 1);
+
+		ConnectionFactory connectionFactory = getKafkaBrokerConnectionFactory();
+		ArrayList<Partition> readPartitions = new ArrayList<Partition>();
+		for (int i = 0; i < 5; i++) {
+			readPartitions.add(new Partition(TEST_TOPIC, i));
+		}
+
+		final KafkaMessageListenerContainer kafkaMessageListenerContainer =
+				new KafkaMessageListenerContainer(connectionFactory,
+						readPartitions.toArray(new Partition[readPartitions.size()]));
+		MetadataStoreOffsetManager offsetManager = new MetadataStoreOffsetManager(connectionFactory);
+		SimpleMetadataStore metadataStore = new SimpleMetadataStore();
+		offsetManager.setMetadataStore(metadataStore);
+		kafkaMessageListenerContainer.setOffsetManager(offsetManager);
+		kafkaMessageListenerContainer.setMaxFetch(100);
+		kafkaMessageListenerContainer.setConcurrency(1);
+		kafkaMessageListenerContainer.setStopTimeout(3000);
+
+		int expectedMessageCount = 1;
+
+
+		KafkaMessageDrivenChannelAdapter kafkaMessageDrivenChannelAdapter =
+				new KafkaMessageDrivenChannelAdapter(kafkaMessageListenerContainer);
+
+		final CountDownLatch latch = new CountDownLatch(expectedMessageCount);
+
+		StringDecoder decoder = new StringDecoder();
+		kafkaMessageDrivenChannelAdapter.setKeyDecoder(decoder);
+		kafkaMessageDrivenChannelAdapter.setPayloadDecoder(decoder);
+		kafkaMessageDrivenChannelAdapter.setBeanFactory(mock(BeanFactory.class));
+		kafkaMessageDrivenChannelAdapter.setOutputChannel(new MessageChannel() {
+			@Override
+			public boolean send(Message<?> message) {
+				try {
+					Thread.sleep(2000L);
+				}
+				catch (InterruptedException e) {
+					e.printStackTrace();
+				}
+				return true;
+			}
+
+			@Override
+			public boolean send(Message<?> message, long timeout) {
+				return send(message);
+			}
+		});
+
+		kafkaMessageDrivenChannelAdapter.setAutoCommitOffset(false);
+		kafkaMessageDrivenChannelAdapter.afterPropertiesSet();
+		kafkaMessageDrivenChannelAdapter.start();
+
+		createStringProducer(NoCompressionCodec$.MODULE$.codec()).send(createMessages(1, TEST_TOPIC));
+
+		long startTime = System.currentTimeMillis();
+		kafkaMessageDrivenChannelAdapter.stop();
+		Assert.assertThat(System.currentTimeMillis() - startTime, Matchers.greaterThan(1500L));
```

Comment:
Actually, that's not correct. Stop() may finish before handleMessage() if it times out.